lib/translate/inquiry.cc

```diff
@@ -0,0 +1,298 @@
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "inquiry.h"
+#include "string.h"
+namespace inquiry {
+// Creates and validates a Inquiry Command struct
+std::optional<scsi_defs::InquiryCommand> raw_cmd_to_scsi_command(
+    absl::Span<const uint32_t> raw_cmd) {
+  if (raw_cmd.empty()) {
+    printf("buffer is empty or nullptr\n");
+    return std::nullopt;
+  }
+
+  uint8_t opcode = raw_cmd[0];
+  if (static_cast<scsi_defs::OpCode>(opcode) != scsi_defs::OpCode::kInquiry) {
+    printf("invalid opcode. expected %ux. got %ux.",
+           static_cast<uint8_t>(scsi_defs::OpCode::kInquiry), opcode);
+    return std::nullopt;
+  }
+
+  // printf("opcode is good, casting\n");
+  // TODO: check invalid parameters in scsi_command
+  return std::make_optional(
+      *reinterpret_cast<const scsi_defs::InquiryCommand*>(&raw_cmd[1]));
+}
+
+// Executes the NVME Identify Controller command
+nvme_defs::IdentifyControllerData nvme_identify_controller() {
+  // TODO
+  return nvme_defs::IdentifyControllerData();
+}
+
+// Executes the NVME Identify Namespace command
+nvme_defs::IdentifyNamespace nvme_identify_namespace() {
+  // TODO
+  return nvme_defs::IdentifyNamespace();
+}
+
+scsi_defs::InquiryData translate_standard_inquiry_response(
+    nvme_defs::IdentifyControllerData identify_controller_data,
+    nvme_defs::IdentifyNamespace identify_namespace_data) {
+  // SCSI Inquiry Standard Result
+  // https://www.nvmexpress.org/wp-content/uploads/NVM-Express-SCSI-Translation-Reference-1_1-Gold.pdf
+  // Section 6.1.1
+  scsi_defs::InquiryData result = scsi_defs::InquiryData();
+  result.peripheral_qualifier = static_cast<scsi_defs::PeripheralQualifier>(0);
+  result.peripheral_device_type =
+      static_cast<scsi_defs::PeripheralDeviceType>(0);
+  result.rmb = 0;
+  result.version = static_cast<scsi_defs::Version>(0x6);
+  result.normaca = 0;
+  result.hisup = 0;
+  result.response_data_format =
+      static_cast<scsi_defs::ResponseDataFormat>(0b10);
+  result.additional_length = 0x1f;
+  result.sccs = 0;
+  result.acc = 0;
+  result.tpgs = static_cast<scsi_defs::TPGS>(0);
+  result.third_party_copy = 0;
+  result.protect = (identify_namespace_data.dps.pit == 0 &&
+                    identify_namespace_data.dps.md_start == 0)
+                       ? 0
+                       : 1;
+  result.encserv = 0;
+  result.multip = 0;
+  result.addr_16 = 0;
+  result.wbus_16 = 0;
+  result.sync = 0;
+  result.cmdque = 1;
+
+  // Shall be set to “NVMe” followed by 4 spaces: “NVMe “
+  char NVME_VENDOR_IDENTIFICATION[9] = "NVMe    ";
```

Comment:
maybe we can extract a constant for this and put it in a `common.h` if other places also need to set this vendor identification